astropy/io/fits/convenience.py

```diff
@@ -482,7 +493,7 @@ def table_to_hdu(table):
             if column.dtype.kind == 'f' and np.allclose(fill_value, 1e20):
                 column.set_fill_value(np.nan)
 
-        table_hdu = BinTableHDU.from_columns(np.array(table.filled()))
+        table_hdu = BinTableHDU.from_columns(np.array(table.filled()), header=hdr)
```

Comment:
Yes, that would be way simpler. I found it difficult to make changes to Column since it happens at the very end and through multiple classes. Maintaining the time state till then was a little tricky.